### ISSUE TITLE
Supervised Contrastive Learning and Feature Fusion for Improved Kinship Verification

### ISSUE THREAD

**Issue by vitalwarley** [vitalwarley/research#80]
Não encontrei código associado, todavia parece simples de reproduzir.

**Comment by vitalwarley** [comment]
_ChatGPT summary_

The paper titled "Supervised Contrastive Learning and Feature Fusion for Improved Kinship Verification" by Nazim Bendib proposes a novel approach to facial kinship verification (FKV) using supervised contrastive learning combined with feature fusion. This methodology aims to enhance the ability to verify kinship relationships by training a model to maximize the similarity between related individuals and minimize it between unrelated individuals.

### Key Points:

1. **Introduction**:
   - FKV involves determining the familial relationship between two facial images and has applications in genealogy, forensic science, social media, and more.
   - Traditional approaches can be divided into shallow models using handcrafted features and deep models using Convolutional Neural Networks (CNNs). However, these approaches often struggle with variations in facial expressions, lighting, pose, and other factors.
   - The paper proposes a new method to address these issues using supervised contrastive learning and feature fusion.

2. **Related Work**:
   - The paper highlights the evolution of FKV research, noting the superiority of deep learning-based methods over handcrafted ones due to their ability to exploit large datasets.
   - Other approaches mentioned include using autoencoders and Generative Adversarial Networks (GANs) to handle age variations between samples.

3. **Method**:

![image](https://github.com/vitalwarley/research/assets/6365065/d74cdb56-89d3-4759-acc7-8aa98f92ed7d)

   - The proposed approach involves two stages:
     1. **Supervised Contrastive Learning**: The encoder network extracts facial representations, which are then mapped into a lower-dimensional space by a projection head. A supervised contrastive loss is applied to maximize the similarity between related pairs and minimize it between unrelated pairs.
     2. **Binary Classification**: A pre-trained encoder is used to extract features, which are then fed into a binary classifier to predict kinship. Feature fusion is employed to enhance classification by combining multiple quadratic combinations of representation vectors.

5. **Batch Sampling**:
   - The paper introduces a batch sampling technique to address the imbalance in the number of photos per individual and the number of individuals per family in the dataset. The technique ensures batches contain distinct families and valid kinship pairs.

6. **Experiments and Results**:

![image](https://github.com/vitalwarley/research/assets/6365065/f3da5fbd-e50c-4c18-aadd-814f64b83374)
![image](https://github.com/vitalwarley/research/assets/6365065/84a2bbf1-d92b-4f2d-bae4-c01f7deacb75)

_Não citam o #26_

   - The method is validated on the Families In the Wild (FIW) dataset, achieving a state-of-the-art accuracy of 81.1%.
   - The paper demonstrates that the proposed method outperforms previous approaches, particularly in handling the age gap between different kinship pairs.
   - Experiments involved both freezing and fine-tuning the pre-trained encoder, with fine-tuning yielding the best results.

8. **Conclusion**:
   - The paper concludes that the proposed supervised contrastive learning approach significantly improves FKV accuracy. Future work will focus on deploying this model on edge devices for real-time kinship verification with enhanced privacy and security.

**Comment by vitalwarley** [comment]
_QA by ChatGPT_

### Method:

1. **Supervised Contrastive Learning**:
   - **What is the role of the encoder \( f(.) \) in the supervised contrastive learning stage?**
     - The encoder \( f(.) \) is responsible for extracting high-dimensional facial representations from the input images. These representations are then used for further processing in the contrastive learning stage.
     
   - **How does the projection head \( g(.) \) contribute to the overall learning process?**
     - The projection head \( g(.) \) maps the high-dimensional representations from the encoder to a lower-dimensional space where the contrastive loss is applied. This helps in learning more discriminative and relevant features by focusing on essential aspects of the data.
     
   - **Can you explain the supervised contrastive loss function used in this approach?**
     - The supervised contrastive loss function aims to maximize the similarity between representations of related individuals and minimize the similarity between unrelated individuals. The loss for a positive pair \((z_i, z_j)\) is calculated using cosine similarity, encouraging the model to learn meaningful relationships between related faces.
     
   - **Why was cosine similarity chosen as the similarity metric for the supervised contrastive loss?**
     - Cosine similarity is chosen because it effectively measures the angle between two vectors, which is useful for determining the similarity between facial representations regardless of their magnitude. This property is beneficial in high-dimensional spaces where the magnitude may not be as important as the direction.
     
   - **What advantages does supervised contrastive learning offer over traditional contrastive learning methods in the context of FKV?**
     - Supervised contrastive learning leverages labeled data to explicitly guide the model in distinguishing between related and unrelated pairs, leading to better discriminative features. This is particularly important in FKV, where the relationships between individuals can be subtle and complex.

2. **Binary Classification**:
   - **How is the binary classifier \( d(.) \) trained in the second stage of the method?**
     - The binary classifier \( d(.) \) is trained on the representations extracted by the pre-trained encoder. It takes pairs of representations as input and predicts whether they belong to related individuals or not, using a feature fusion approach to enhance its decision-making process.
     
   - **What are the key features extracted by the pre-trained encoder \( f(.) \) used in the binary classification stage?**
     - The key features extracted by the encoder include various discriminative facial attributes such as shapes, distances, and textures that are essential for determining kinship relationships. These features are used by the binary classifier to make accurate predictions.
     
   - **How does feature fusion improve the performance of the binary classifier in predicting kinship?**
     - Feature fusion improves performance by combining multiple quadratic combinations of the representation vectors, allowing the classifier to capture more complex interactions and relationships between features. This leads to more accurate kinship verification.

3. **Overall Methodology**:
   - **How does the proposed method address the common issues faced by traditional FKV approaches?**
     - The proposed method addresses issues such as sensitivity to facial variations and suboptimal feature representation by using supervised contrastive learning to learn more robust and discriminative features. The batch sampling technique also ensures balanced and representative training data.
     
   - **What are the main contributions of this paper to the field of facial kinship verification?**
     - The main contributions include defining FKV as a contrastive learning problem, introducing a novel batch sampling method, and achieving state-of-the-art accuracy on the FIW dataset through a combination of supervised contrastive learning and feature fusion.
     
   - **Can you describe the architecture of the neural network encoder and the projection head used in this study?**
     - The encoder is based on ArcFace100, a deep CNN architecture known for its effectiveness in facial recognition tasks. The projection head is a 2-layer MLP that maps the high-dimensional facial representations into a 128-dimensional space for contrastive learning.

### Batch Sampling:

1. **General Approach**:
   - **What challenges in the FIW dataset necessitate the use of a specialized batch sampling technique?**
     - The FIW dataset is imbalanced, with varying numbers of photos per individual and different numbers of individuals per family. This imbalance can lead to biased training and poor generalization, necessitating a specialized batch sampling technique to ensure fair and effective training.
     
   - **How does the batch sampling technique ensure that the kinship constraints are not violated?**
     - The batch sampling technique ensures kinship constraints are not violated by carefully selecting pairs of images that represent valid kinship relationships. It avoids sampling non-kin pairs from the same family and maintains a balanced representation of different families in each batch.

2. **Implementation Details**:
   - **What criteria are used to select images with the minimum counter for each individual in a batch?**
     - Images with the minimum counter are selected to ensure that each image is used approximately equally throughout the training process. This prevents overfitting to specific images and promotes a more balanced and comprehensive learning experience.
     
   - **How does the batch sampling algorithm handle the imbalance in the number of photos per individual and the number of individuals per family?**
     - The algorithm handles imbalance by shuffling the list of pairs, removing duplicate families within each batch, and replacing them with distinct families. It selects images with the minimum counter to balance the representation of individuals and families across batches.
     
   - **Can you explain the steps involved in replacing duplicate families within a batch to ensure distinct family representation?**
     - When a batch contains duplicate families, the algorithm replaces these duplicates with other pairs from distinct families. This is done to avoid having negative pairs from the same family, ensuring that the model does not get confused by invalid kinship relationships.

3. **Impact on Training**:
   - **How does the batch sampling technique contribute to the overall performance and accuracy of the model?**
     - The batch sampling technique contributes to overall performance by ensuring that the training data is balanced and representative. This leads to more robust and generalized learning, reducing overfitting and improving accuracy.
     
   - **In what ways does the proposed batch sampling technique differ from conventional batch sampling methods used in deep learning?**
     - The proposed batch sampling technique differs by specifically addressing the unique challenges of kinship verification, such as the need to avoid invalid kinship pairs and ensure balanced representation of families. Conventional batch sampling methods do not typically consider these constraints, leading to suboptimal performance in FKV tasks.

**Comment by vitalwarley** [comment]
Sure, here's a PyTorch implementation of the batch sampler algorithm described in the paper:

```python
import random
from collections import defaultdict

class KinshipBatchSampler:
    def __init__(self, relationships, batch_size):
        self.relationships = relationships
        self.batch_size = batch_size
        self.image_counters = defaultdict(int)
        self._shuffle_relationships()

    def _shuffle_relationships(self):
        random.shuffle(self.relationships)

    def _get_image_with_min_count(self, person_images):
        min_count_image = min(person_images, key=lambda img: self.image_counters[img])
        return min_count_image

    def _replace_duplicates(self, sub_batch):
        unique_families = set()
        for pair in sub_batch:
            unique_families.add(pair[2])  # Assuming the third element is the family identifier
        
        while len(unique_families) < len(sub_batch):
            for i in range(len(sub_batch)):
                if sub_batch[i][2] in unique_families:
                    continue
                replacement_pair = random.choice(self.relationships)
                if replacement_pair[2] not in unique_families:
                    sub_batch[i] = replacement_pair
                    unique_families.add(replacement_pair[2])
        return sub_batch

    def __iter__(self):
        for i in range(0, len(self.relationships), self.batch_size):
            sub_batch = self.relationships[i:i + self.batch_size]
            if len(sub_batch) < self.batch_size:
                continue

            sub_batch = self._replace_duplicates(sub_batch)
            batch = []

            for pair in sub_batch:
                per1, per2, fam = pair
                img1 = self._get_image_with_min_count(per1)
                img2 = self._get_image_with_min_count(per2)
                self.image_counters[img1] += 1
                self.image_counters[img2] += 1
                batch.append((img1, img2))

            yield batch

# Example usage:
# relationships = [
#     (['img1_1', 'img1_2'], ['img2_1', 'img2_2'], 'fam1'),
#     (['img3_1', 'img3_2'], ['img4_1', 'img4_2'], 'fam2'),
#     ...
# ]
# batch_size = 32
# sampler = KinshipBatchSampler(relationships, batch_size)
# for batch in sampler:
#     print(batch)
```

### Explanation:

1. **Initialization (`__init__`)**:
   - `relationships`: A list of tuples where each tuple contains images of two persons and their family identifier.
   - `batch_size`: The size of each batch.
   - `image_counters`: A dictionary to keep track of how many times each image has been used.
   - `_shuffle_relationships()`: Shuffles the list of relationships.

2. **Helper Methods**:
   - `_shuffle_relationships()`: Shuffles the `relationships` list to ensure randomness.
   - `_get_image_with_min_count(person_images)`: Selects the image with the minimum count for a given person.
   - `_replace_duplicates(sub_batch)`: Ensures that each sub-batch contains unique families by replacing duplicates with new pairs from different families.

3. **Iterator (`__iter__`)**:
   - Iterates over the relationships in chunks of `batch_size`.
   - Calls `_replace_duplicates(sub_batch)` to ensure each batch has unique families.
   - For each pair in the batch, selects images with the minimum count, updates the counter, and appends to the batch.

### Example Usage:
- You can use this batch sampler with a dataset of relationships by initializing the `KinshipBatchSampler` with the relationships list and desired batch size.
- Iterate over the sampler to get batches of image pairs for training.

This implementation ensures balanced and representative training batches while adhering to kinship constraints.

![image](https://github.com/vitalwarley/research/assets/6365065/6e10a4d4-1e16-4ad9-b94e-57f3a57bf679)

Here's a comparison between the provided algorithm and my initial implementation:

### Provided Algorithm:
1. **Input**:
   - `relationships`: A list of image pairs and their family relationships.
   - `batch_size`: The size of each batch.

2. **Main Loop**:
   - Iterate over `relationships` in chunks of `batch_size`.
   - For each chunk (`sub`):
     - Check if the families in `sub` are distinct. If not, replace duplicates.
     - Initialize an empty `batch`.
     - For each pair in `sub`:
       - Extract individuals `per1` and `per2`, and the family identifier `fam`.
       - Select images `img1` and `img2` with the minimum count for `per1` and `per2`.
       - Increment the count for `img1` and `img2`.
       - Append the pair `(img1, img2)` to `batch`.
     - Yield the `batch`.

### My Implementation:
```python
import random
from collections import defaultdict

class KinshipBatchSampler:
    def __init__(self, relationships, batch_size):
        self.relationships = relationships
        self.batch_size = batch_size
        self.image_counters = defaultdict(int)
        self._shuffle_relationships()

    def _shuffle_relationships(self):
        random.shuffle(self.relationships)

    def _get_image_with_min_count(self, person_images):
        min_count_image = min(person_images, key=lambda img: self.image_counters[img])
        return min_count_image

    def _replace_duplicates(self, sub_batch):
        unique_families = set()
        for pair in sub_batch:
            unique_families.add(pair[2])  # Assuming the third element is the family identifier
        
        while len(unique_families) < len(sub_batch):
            for i in range(len(sub_batch)):
                if sub_batch[i][2] in unique_families:
                    continue
                replacement_pair = random.choice(self.relationships)
                if replacement_pair[2] not in unique_families:
                    sub_batch[i] = replacement_pair
                    unique_families.add(replacement_pair[2])
        return sub_batch

    def __iter__(self):
        for i in range(0, len(self.relationships), self.batch_size):
            sub_batch = self.relationships[i:i + self.batch_size]
            if len(sub_batch) < self.batch_size:
                continue

            sub_batch = self._replace_duplicates(sub_batch)
            batch = []

            for pair in sub_batch:
                per1, per2, fam = pair
                img1 = self._get_image_with_min_count(per1)
                img2 = self._get_image_with_min_count(per2)
                self.image_counters[img1] += 1
                self.image_counters[img2] += 1
                batch.append((img1, img2))

            yield batch

# Example usage:
# relationships = [
#     (['img1_1', 'img1_2'], ['img2_1', 'img2_2'], 'fam1'),
#     (['img3_1', 'img3_2'], ['img4_1', 'img4_2'], 'fam2'),
#     ...
# ]
# batch_size = 32
# sampler = KinshipBatchSampler(relationships, batch_size)
# for batch in sampler:
#     print(batch)
```

### Comparison:

1. **Initialization**:
   - Both versions initialize the relationships and batch size. My implementation also initializes a counter for image usage and shuffles the relationships.

2. **Shuffle Relationships**:
   - My implementation includes a `_shuffle_relationships` method to ensure randomness. The provided algorithm doesn't specify this step, but it is generally a good practice.

3. **Replace Duplicates**:
   - Both versions include a mechanism to replace duplicate families within a batch. My implementation does this through the `_replace_duplicates` method, which ensures each batch has distinct families by shuffling and replacing pairs if needed.

4. **Image Selection**:
   - Both versions select images with the minimum count for each person to ensure balanced usage. My implementation uses `_get_image_with_min_count` for this purpose.

5. **Batch Creation**:
   - Both versions iterate through the pairs in the sub-batch, select the appropriate images, increment their usage count, and append the pair to the batch.

6. **Yield Batch**:
   - Both versions yield the constructed batch at the end of the iteration.

### Conclusion:
The provided algorithm and my implementation are very similar in structure and logic. My implementation includes some additional details, such as shuffling the relationships and explicit method definitions, which align with the steps in the provided algorithm. Both approaches ensure balanced and representative training batches while adhering to kinship constraints.
